cluster/src/main/java/io/scalecube/cluster/ClusterImpl.java

```diff
@@ -1,243 +1,239 @@
-
-          // stop transport
-          transport.stop(transportStoppedFuture);
-
-          LOGGER.info("Cluster member has shut down... {}", membership.member());
-        });
-    return transportStoppedFuture;
-  }
-
-  @Nonnull
-  @Override
-  public NetworkEmulator networkEmulator() {
-    return transport.networkEmulator();
-  }
-
-  @Override
-  public boolean isShutdown() {
-    return this.transport.isStopped(); // since transport is the last component stopped on shutdown
-  }
-}
+package io.scalecube.cluster;
+
+import static io.scalecube.Preconditions.checkNotNull;
+import static io.scalecube.cluster.fdetector.FailureDetectorImpl.PING;
+import static io.scalecube.cluster.fdetector.FailureDetectorImpl.PING_ACK;
+import static io.scalecube.cluster.fdetector.FailureDetectorImpl.PING_REQ;
+import static io.scalecube.cluster.gossip.GossipProtocolImpl.GOSSIP_REQ;
+import static io.scalecube.cluster.membership.MembershipProtocolImpl.MEMBERSHIP_GOSSIP;
+import static io.scalecube.cluster.membership.MembershipProtocolImpl.SYNC;
+import static io.scalecube.cluster.membership.MembershipProtocolImpl.SYNC_ACK;
+
+import io.scalecube.cluster.fdetector.FailureDetectorImpl;
+import io.scalecube.cluster.gossip.GossipProtocolImpl;
+import io.scalecube.cluster.membership.MembershipEvent;
+import io.scalecube.cluster.membership.MembershipProtocolImpl;
+import io.scalecube.transport.Address;
+import io.scalecube.transport.Message;
+import io.scalecube.transport.NetworkEmulator;
+import io.scalecube.transport.Transport;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import reactor.core.publisher.Flux;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Cluster implementation.
+ * 
+ */
+final class ClusterImpl implements Cluster {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ClusterImpl.class);
+
+  private static final Set<String> SYSTEM_MESSAGES =
+      Collections.unmodifiableSet(Stream.of(PING, PING_REQ, PING_ACK, SYNC, SYNC_ACK, GOSSIP_REQ)
+          .collect(Collectors.toSet()));
+
+  private static final Set<String> SYSTEM_GOSSIPS = Collections.singleton(MEMBERSHIP_GOSSIP);
+
+  private final ClusterConfig config;
+
+  private final ConcurrentMap<String, Member> members = new ConcurrentHashMap<>();
+  private final ConcurrentMap<Address, String> memberAddressIndex = new ConcurrentHashMap<>();
+
+  // Cluster components
+  private Transport transport;
+  private FailureDetectorImpl failureDetector;
+  private GossipProtocolImpl gossip;
+  private MembershipProtocolImpl membership;
+
+  private Flux<Message> messageObservable;
+  private Flux<Message> gossipObservable;
+

```

Comment:
Why keep those variables: `messageObservable` and `gossipObservable` . IMHO the gain we get is super-minor, that doesn't deserves decalration of two fields.

Reply:
@artem-v  its keeping same as was before this change you can suggest alternative to this pull request?

Reply:
@ronenhamias 

I looked more precisely and it appears like we assign those field variables right at the lambda, at 
```java
join0() {
transportFuture = Transport.bind()
transportFuture.thenCompose(boundTransport -> {
  messageObservable = ...
  gossipObservable = ...
})
}
```
It looks like we have to start listening on cluster protocol port  before doing any aggregation phases (which is totally weird IMHO)